src/coreclr/jit/importer.cpp

```diff
@@ -5469,18 +5469,18 @@ GenTree* Compiler::impCastClassOrIsInstToTree(
         }
         else if (isCastClass)
         {
-            // Jit can only inline expand CHKCASTCLASS and CHKCASTARRAY helpers.
-            canExpandInline = (helper == CORINFO_HELP_CHKCASTCLASS) || (helper == CORINFO_HELP_CHKCASTARRAY);
+            // Jit can always inline expand CHKCASTCLASS
+            canExpandInline = (helper == CORINFO_HELP_CHKCASTCLASS);
 
             // For ChkCastAny we ignore cases where the class is known to be abstract or is an interface.
             if (helper == CORINFO_HELP_CHKCASTANY)
             {
                 const bool isAbstract = (info.compCompHnd->getClassAttribs(pResolvedToken->hClass) &
-                                         (CORINFO_FLG_INTERFACE | CORINFO_FLG_ABSTRACT)) != 0;
+                    (CORINFO_FLG_INTERFACE | CORINFO_FLG_ABSTRACT)) != 0;
                 canExpandInline = !isAbstract;
             }
         }
-        else if ((helper == CORINFO_HELP_ISINSTANCEOFCLASS) || (helper == CORINFO_HELP_ISINSTANCEOFARRAY))
+        else if ((helper == CORINFO_HELP_ISINSTANCEOFCLASS) || (helper == CORINFO_HELP_CHKCASTANY))
         {
             // If the class is exact, the jit can expand the IsInst check inline.
```

Comment:
I think this should do isAbstract check like the above.

Reply:
As I understand `isClassExact` implies that the type is a nonvariant final. The interfaces that may make here will be either variant or use `CORINFO_HELP_CHKCASTINTERFACE`.

Reply:
That is - unless we want to do inline expansion for throwing and nonthrowing path under the same conditions. I kind of assumed the difference is intentional.

Reply:
I think the idea is that we want to inline the exact match if it has a chance of succeeding.

@EgorBo Do you have any additional insights about this logic?

Reply:
Also it looks like inline expansion of `isinst` generally does not call the helper, just produces `null` when types do not match. 
https://github.com/dotnet/runtime/blob/fff52666cf669946254b07764e464f75f0b69c38/src/coreclr/jit/importer.cpp#L5637

If that is the case, then "inlining" in abstract case is still the right thing to do - no need to call the helper. The intersection of `isClassExact` and `isAbstract` could probably just optimize the whole deal into `null`. Although it might be too rare to worry about.

I am also wondering how all this works with COM, type equivalency, ICastable and other similar things. How can we avoid calling the helper if the object to which `isinst` is applied could be one of those cases? 
The more I look at this the more confusing it gets...

Reply:
> I am also wondering how all this works with COM, type equivalency, ICastable and other similar thing

These things are handled by the slow paths of the helpers. It means that the ininst negative case is optimizable only when we can be sure that none of these rare cases are possible.

Reply:
> I think this should do isAbstract check like the above.

My understanding that we try to avoid expansion for `foo is T` if `T` is not exact, but we currently do that for `(T)foo` because the latter, presumably, has less options and InvalidCastException is not expected to be popular.

We can, probably, enable it for the former as well if we collect some numbers on how often this check is taken, right?
For JIT, I think, we'll eventually enable PGO for casts by default.

Reply:
E.g.:
```cs
public class ClassA { }
public class ClassB : ClassA { }

class Program
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    bool IsClassA(object obj) => obj is ClassA;

    static void Main()
    {
        Program p = new();
        for (int i = 0; i < 100; i++)
        {
            p.IsClassA(new ClassB());
            Thread.Sleep(16);
        }
    }
}
```
Current codegen for `IsClassA`:
```asm
; Assembly listing for method Program:IsClassA(System.Object):bool:this (Tier1)
       sub      rsp, 40
       mov      rcx, 0x7FFC89533648      ; ClassA
       call     [CORINFO_HELP_ISINSTANCEOFCLASS]
       test     rax, rax
       setne    al
       movzx    rax, al
       add      rsp, 40
       ret 
```
(and if we partially expand it for `ClassA` it will be a pure overhead)
Now, codegen with `DOTNET_JitProfileCasts=1`:
```asm
; Assembly listing for method Program:IsClassA(System.Object):bool:this (Tier1)
       sub      rsp, 40
       mov      rax, rdx
       test     rax, rax
       je       SHORT G_M61590_IG05
       mov      rcx, 0x7FFC89503328      ; ClassB
       cmp      qword ptr [rax], rcx
       je       SHORT G_M61590_IG05
       mov      rcx, 0x7FFC895031E0
       call     [CORINFO_HELP_ISINSTANCEOFCLASS]
G_M61590_IG05:
       test     rax, rax
       setne    al
       movzx    rax, al
       add      rsp, 40
       ret
```

Reply:
I am still confused at what makes the `IsInst` optimization correct for types that have special casting semantics - like `ICastable`, COM or something with type equivalence. 

`impIsClassExact` does not check for these cases. It only checks for "final" + "nonvariant" case (in array case, the element is checked)

If a "special" casting is always to an interface, then, yes, we will never optimize "special" scenarios as they would either use interface-specific helpers (`CHKCASTINTERFACE`/`ISINSTANCEOFINTERFACE`) or will be filtered out due to variance. 
Am not sure it is always to an interface though.  

Basically if we have 
```cs
bool converts = obj is SomeSealedClass;
```

and `obj` happens to convert to `SomeSealedClass` via ICastable or type equivalence, would we optimize this case incorrectly into a type comparison?

Reply:
Are you able to construct a repro that shows that this optimization is incorrect for these corner cases?

I think it is likely that you are right that this optimization is incorrect for some corner cases. The `impIsClassExact ` pattern where the JIT tries to figure some kind of non-trivial type system property by combining type flags on its own is bug prone. It is better to have these non-trivial type system properties to be computed by dedicated JIT/EE interface APIs.

Reply:
Also, it seems that some of these buggy corner cases could have been introduced by deleting `CORINFO_HELP_ISINSTANCEOFARRAY`/`CORINFO_HELP_CHKCASTARRAY` helpers. We should get closure on this before merging this PR.

Reply:
Yes, that is my worry - we may have subtle holes here and I may be making things worse. Even though tests are passing.

It is possible to check that the destination type is an array and match the behavior as if array-specific helpers are still there, but need to be more sure that the current behavior is correct.  

Checked casts are ok, since we call helpers on nonmatch, but isinst is a concern.

Reply:
I looked at the cases when an object may cast to/from something that is not in its inheritance. We have 3 cases:
- `ICastable`   - not a concern, since the destination type must be an interface (so it will use different helpers)
- `IDynamicCastable` - same as above, not a concern.
- Type Equivalence  - it seems to be possible, at least in theory, to cast to an otherwise unrelated and sealed type via type equivalence. 

**I think types that participate in type equivalence should be rejected by `impIsClassExact` per the helper's intent.** 
I've added a commit that implements that - https://github.com/dotnet/runtime/pull/89548/commits/3a38456b92864e3af45727a4c285d7a3fed44231

I, however, could not figure how to construct an example that would demonstrate a problem. While I see how type equivalence works in the type system, I do not completely understand how such type are created by programs.

I have tried the following, but it did not result in types compatible via equivalence. Perhaps something is missing?

```cs
using System.Runtime.InteropServices;

[assembly:ImportedFromTypeLib("qq")]

namespace ConsoleApp33
{
    [TypeIdentifier("a", "b")]
    struct T1
    {
    }

    [TypeIdentifier("a", "b")]
    struct T2
    {
    }

    internal class Program
    {
        static void Main(string[] args)
        {
            object o = new T1();
            Console.WriteLine(o is T2);
            Console.WriteLine(typeof(T1).IsEquivalentTo(typeof(T2)));

            T2 c = (T2) o;
            Console.WriteLine (c.ToString ());
        }
    }
}
```

Reply:
> I have tried the following, but it did not result in types compatible via equivalence. Perhaps something is missing?

The type names need to match. You need a helper assembly with the equivalent type to make this work. For example, look for `EquivalentValueType` under `src\libraries\System.Reflection\tests`.

Reply:
It appears to be possible to add a scenario to EquivalentValueType test that is sensitive to the `CORINFO_FLG_TYPE_EQUIVALENCE` fix.  

Another important part is to force `AggressiveOptimization` to make sure that JIT optimizes the cast.